activerecord/lib/arel/visitors/to_sql.rb

```diff
@@ -24,13 +24,19 @@ def visit_Arel_Nodes_DeleteStatement(o, collector)
 
           if has_join_sources?(o)
             collector << "DELETE "
+
             visit o.relation.left, collector
             collector << " FROM "
           else
-            collector << "DELETE FROM "
+            collector << "DELETE "
+            if o.instance_values["relation"].instance_values["table_alias"] && collector.instance_values["delegate"]
```

Comment:
This code can't live here. The visitors should not have to inspect the collector. We should make sure the AST being build is valid, not accepting an invalid AST and trying to figure out what state it is by inspecting the buggers we have.

Reply:
Which delete statement is this trying to generate? I can't see MySQL, PostgreSQL or Sqlite accepting anything before the `FROM`